### ISSUE TITLE
Autosize of windows is not work properly on different DPI scalings

### ISSUE THREAD

**Issue by Alexpux** [cnjinhao/nana#677]
On Win10 when you change scaling factor in settings layouts autosize not work properly. From demo projects here is inputbox pictures for 100% and 125% scaling:
![right_layout_size](https://github.com/cnjinhao/nana/assets/1898153/88f605ae-dfff-4773-8f52-087e752f11fe)
![wrong_layout_size](https://github.com/cnjinhao/nana/assets/1898153/82ea6188-c454-441d-bdf3-6dee274f3ca8)

**Comment by cnjinhao** [comment]
Thank you for reporting this issue. I will check it.

**Comment by edgarbernal5** [comment]
Hi! I don't know if it is the same issue but when I moved the form created in a 4k monitor to a HD monitor the title bar (caption, minimize, maximize, etc) won't scale the title bar correctly.

4K monitor
![nana test 4k](https://github.com/cnjinhao/nana/assets/1410581/0e6994fc-c876-4050-b469-0d10ebb30fb8)

HD monitor
![nana test hd](https://github.com/cnjinhao/nana/assets/1410581/7c73fe96-5652-4696-9fae-80f615ba3f57)

**Comment by edgarbernal5** [comment]
Also I managed to scale menubar with different DPI by getting widget size of it. Here is the code within source/gui/widgets/menubar.cpp

```
std::size_t find(const ::nana::point& pos)
{
  // changed this line:
  if ((2 <= pos.x) && (2 <= pos.y) && (pos.y < 25))
  
  //with this:
  auto menubar_size = widget_ptr->size();
  if ((2 <= pos.x) && (2 <= pos.y) && (pos.y < menubar_size.height))
```
....

Also on refresh method

```
void trigger::refresh(graph_reference graph)
{
....
  auto menubar_size = ess_->widget_ptr->size();
  nana::point item_pos(2, 2);
  nana::size item_s(0, menubar_size.height - 2);
...
```

What do you think? Thanks in advance.
There is still a bug with menu popoups

**Comment by cnjinhao** [comment]
@edgarbernal5 a PR is welcomed.

**Comment by edgarbernal5** [comment]
Hi @cnjinhao, I want to contribute, I have a commit ready to be pushed but I think I don't have right access:

 git push --set-upstream origin dpi_scale_issue
remote: Permission to cnjinhao/nana.git denied to edgarbernal5.
fatal: unable to access 'https://github.com/cnjinhao/nana.git/': The requested URL returned error: 403

**Comment by qPCR4vir** [comment]
@edgarbernal5 make your push into your own fork of nana https://github.com/edgarbernal5/nana, and from there open a pull request to the original fork.

**Comment by edgarbernal5** [comment]
In order to create the pull request, should I create a branch in my own fork with all in? @qPCR4vir

**Comment by edgarbernal5** [comment]
I think I know the answer 😀

**Comment by edgarbernal5** [comment]
Pull request: #690.
Any feedback is welcomed.

**Comment by cnjinhao** [comment]
> Pull request: #690.
> 
> Any feedback is welcomed.

Thanks, the pull request has been merged.

**Comment by qPCR4vir** [comment]
> Pull request: #690. Any feedback is welcomed.

Hi ! It looks like a great job. I hope you (@edgarbernal5 ) will keep doing soo.

But in your Pull Request you (@edgarbernal5 ) made the same mistake I made some times: there are a lot of unrelated commits that make it very hard to follow what you did. I'm sorry that I did not review that, and only now report it. To fix this I propose that you both go to the previous commit:  7c73095f26e4a7e2b29af6097e692ba7e2cd6d71  "fix crash and rectangle errors in pixel_buffer" and make a hard reset of the branch to that. @cnjinhao will need to rewirte that to the repo (force push that branch again). After that one (@edgarbernal5 ) need to do a merge , but using a 'cherry pick' and include only the commit fe7536004baadb8e3ace90039d8c560830c8411e "Fixed dpi scale for menu, menubar and combox". In a new Pull request verify only that commit was made. @cnjinhao  and merged and made a new push. This will re-write history. A clean history. There are many ways to do this, just verify before push. If you want I can try to do all of that by myself. We all learn by doing...

![image](https://github.com/cnjinhao/nana/assets/1911982/f8db6a03-561c-473c-b2ad-0300a8e7eb92)

**Comment by edgarbernal5** [comment]
My bad, it was my fault! I want to fix this but I don't know how to do it right! We can do it together if you want @qPCR4vir 

I got the following message:

git reset --hard 7c73095f26e4a7e2b29af6097e692ba7e2cd6d71
fatal: Could not parse object '7c73095f26e4a7e2b29af6097e692ba7e2cd6d71'.

or doing this:

git checkout 7c73095f26e4a7e2b29af6097e692ba7e2cd6d71
fatal: reference is not a tree: 7c73095f26e4a7e2b29af6097e692ba7e2cd6d71

Probably I am missing something.
By the way I want to do the same thing in my fork, to clean the mess, because this project is awesome and want to continue developing and contributing not only with bug fixing but also with creating new widgets.

**Comment by qPCR4vir** [comment]
ok... I did what I could.. I hope it is not too bad now. 
@cnjinhao and @edgarbernal5 will need to rename your local branch develop-1.8 and pull it again from this repo.

**Comment by edgarbernal5** [comment]
Thanks a million! You saved me!

**Comment by qPCR4vir** [comment]
still a lot of problems like this: just set it to 300% for easy detection of problems

**Comment by edgarbernal5** [comment]
@qPCR4vir Sure, no problem. I could take a look. Would you know what are the repro steps?

> still a lot of problems like this: just set it to 300% for easy detection of problems
Do you mean changing scale and layout on display on Windows Settings?

**Comment by qPCR4vir** [comment]
> On Win10 when you change scaling factor in settings layouts autosize not work properly. From demo projects here is `inputbox `pictures for 100%, 125% and 150% scaling:

Yes, with 300% it is easy to see....

**Comment by qPCR4vir** [comment]
Unfortunately, the listbox widget is also affected.

**Comment by edgarbernal5** [comment]
> The constructor of `inputbox_window` in `nana\source\gui\msgbox.cpp` end with: `move(api::make_center(this->owner(), description_size.width, height));` And both the width and the height are bad (too small) by scaling other than 100%.

Okey, got it. Maybe we need to do the dpi scale by calling this:
`move(api::make_center(this->owner(), platform_abstraction::dpi_scale(handle(), description_size.width), platform_abstraction::dpi_scale(handle(), height)));`

**Comment by qPCR4vir** [comment]
Well... I wanted to help with this, but I'm afraid I don't understand why we are doing this:
when I want to set a scale, I go to Windows settings and change the scale for that monitor, and ALL programs just rescale the ENTIRE drawing. Why are we trying to do that separately in nana? I don't expect and don't want any program to rewrite my settings and try to change the scale on its own... or... what am I missing here?
What are we trying to do? If your operating system has no way to rescale a high dpi monitor... well, you need a new operating system. But how will a program guess what the user wants?
I use a 4k monitor (actually a big and cheap Samsung TV), and set 100% in windows setting: all OK, inlcusive nana.
But if I set 200% - all program are now big, but nana only partially. Why? 
Normally I use 100%... but what if one day I want to do a presentation at 200%.. and all program will nicely scale, but not my program with nana? 
What am I missing here? At some point did operating systems offer non-scaling, making high dpi monitors useless? And nana 'fixed' that?

**Comment by qPCR4vir** [comment]
hmm, #640...

**Comment by qPCR4vir** [comment]
OK, I suppose that the goal was to "amplify" extreme small fonts and elements to make it readable for dpi > 96.
I will try to 'fix' all of this by preventing scaling when the used dpi is less than 96 and scaling when it is >96.

**Comment by qPCR4vir** [comment]
Why for font it is `dpi / 72` and for other use `dpi / 96` ?
hmmm, found: https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-logfontw

**Comment by qPCR4vir** [comment]
testing: https://github.com/qPCR4vir/nana/tree/dpi

**Comment by qPCR4vir** [comment]
hmm, no. Windows internally reinterpret all the coordinates, scaling for that % from the windows setting and report the new dpi already "scaled", but your program don't see that - it continue to to use the same old coordinates. We need to know when it was set for a physical monitor or for an 'scaled' monitor.

**Comment by qPCR4vir** [comment]
Sad... I'm not doing any progress. We need a lot of discipline to fix this.

**Comment by edgarbernal5** [comment]
Maybe it is a good idea if we readjust size and position of all widget of root widgets upon WM_DPICHANGED message arrives. What do you guys think about it? @cnjinhao @qPCR4vir

**Comment by qPCR4vir** [comment]
https://learn.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows

**Comment by qPCR4vir** [comment]
OK... as per this issue, we want to adapt to a dpi change: we need DPI_AWARENESS_PER_MONITOR_AWARE (= 2, version 2)
We need to use [GetDpiForWindow](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getdpiforwindow)
It will not work in wind7. 
`int scaled_x = x_opt4_96dpi * dpi / USER_DEFAULT_SCREEN_DPI;   `
USER_DEFAULT_SCREEN_DPI  = 96

**Comment by qPCR4vir** [comment]
please see: #697

**Comment by qPCR4vir** [comment]
go to: #698

**Comment by cnjinhao** [comment]
The current design is not easy for resizing window of DPI change. We need to design an adaptive coordinating system for the feature, it's complicated. IMO, we can add the event of DPI change to let user control the resize.
